executor/analyze.go

```diff
@@ -381,3 +401,444 @@ func (e *AnalyzeColumnsExec) buildStats() (hists []*statistics.Histogram, cms []
+			}
+		}
+	}()
+	var snapshot kv.Snapshot
+	snapshot, *err = e.ctx.GetStore().(tikv.Storage).GetSnapshot(kv.MaxVersion)
+	if *err != nil {
+		*err = errors.Trace(*err)
+		return
+	}
+	for {
+		task, ok := <-e.tasks
+		if !ok {
+			return
+		}
+
+		var tableID, minRowID, maxRowID int64
+		startKey, endKey := task.Location.StartKey, task.Location.EndKey
+		tableID, minRowID, *err = tablecodec.DecodeRecordKey(startKey)
+		if *err != nil {
+			*err = errors.Trace(*err)
+			return
+		}
+		_, maxRowID, *err = tablecodec.DecodeRecordKey(endKey)
+		if *err != nil {
+			*err = errors.Trace(*err)
+			return
+		}
+
+		keys := make([]kv.Key, 0, task.SampSize)
+		for i := 0; i < int(task.SampSize); i++ {
+			randKey := rand.Int63n(maxRowID-minRowID) + minRowID
+			keys = append(keys, tablecodec.EncodeRowKeyWithHandle(tableID, randKey))
+		}
+
+		var kvMap map[string][]byte
+		kvMap, *err = snapshot.BatchGet(keys)
+		if *err != nil {
+			*err = errors.Trace(*err)
+			return
+		}
+
+		*err = e.handleBatchGetResponse(&kvMap, collectors)
+		if *err != nil {
+			*err = errors.Trace(*err)
+			return
+		}
+	}
+}
+
+func (e *AnalyzeFastExec) runTasks() ([]*statistics.Histogram, []*statistics.CMSketch, error) {
+	errs := make([]error, e.concurrency)
+	hasPKInfo := 0
+	if e.pkInfo != nil {
+		hasPKInfo = 1
+	}
+	hasIdxInfo := len(e.idxsInfo)
+	// collect column samples and primary key samples and index samples.
+	length := len(e.colsInfo) + hasPKInfo + hasIdxInfo
+	collectors := make([]*statistics.SampleCollector, length)
+	for i := range collectors {
+		collectors[i] = &statistics.SampleCollector{
+			IsMerger:      true,
+			FMSketch:      statistics.NewFMSketch(maxSketchSize),
+			MaxSampleSize: maxSampleSize,
+			CMSketch:      statistics.NewCMSketch(defaultCMSketchDepth, defaultCMSketchWidth),
+		}
+	}
+
+	e.wg.Add(e.concurrency)
+	for i := 0; i < e.concurrency; i++ {
+		bo := tikv.NewBackoffer(context.Background(), 500)
+		go e.handleSampTasks(bo, &errs[i], collectors)
+	}
+	e.wg.Wait()
+	for _, err := range errs {
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+	}
+
+	var err error
+	bo := tikv.NewBackoffer(context.Background(), 500)
+	err = e.handleScanTasks(bo, collectors)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	hists, cms := make([]*statistics.Histogram, e.concurrency), make([]*statistics.CMSketch, e.concurrency)
+	for i := 0; i < length; i++ {
+		if i == 0 && hasPKInfo > 0 {
+			hists[i], err = statistics.BuildColumn(e.ctx, int64(e.maxNumBuckets), e.pkInfo.ID, collectors[i], &e.pkInfo.FieldType)
+		} else {
+			hists[i], err = statistics.BuildColumn(e.ctx, int64(e.maxNumBuckets), e.colsInfo[i+hasPKInfo].ID, collectors[i], &e.colsInfo[i+hasPKInfo].FieldType)
+		}
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		cms[i] = collectors[i].CMSketch
+	}
+	return hists, cms, nil
+}
+
+func (e *AnalyzeFastExec) buildStats() (hists []*statistics.Histogram, cms []*statistics.CMSketch, err error) {
+	for {
```

Comment:
We need to limit the retry count.

Reply:
OK, is it a constant or a session variable?

Reply:
Currently a constant is fine.